### PR TITLE
Allow . as a relative path

### PR DESCRIPTION
**Pull request by harendra-kumar**
The discussion thread that I started for empty path finally boils down to this. A quick look at the changes in the tests will give you an idea about what it means to the library user. Let me know what you think about this. These are all the changes needed except for some doc tweaks if this gets accepted.

It essentially allows us to compose "." with other relative paths:

> > toFilePath $ $(mkRelDir "./") </> $(mkRelDir "./")
> > "./"
> > toFilePath $ $(mkRelDir "./") </> $(mkRelDir "x")
> > "x/"
> > toFilePath $ $(mkRelDir "x") </> $(mkRelDir ".")
> > "x/"

Also dirname "." is "." itself:

> > dirname $(mkRelDir ".") == $(mkRelDir ".")
> > True

### CODE REVIEW

File: src/Path.hs [hunk 1]
```diff
@@ -259,8 +265,13 @@ mkRelFile s =
...
 -- slash, you can use 'System.FilePath.dropTrailingPathSeparator' from
 -- the filepath package.
-toFilePath :: Path b t -> FilePath
-toFilePath (Path l) = l
+toFilePath :: (Typeable b, Typeable t) => Path b t -> FilePath
+toFilePath l =
+  case l of
+    Path [] -> case cast l of
+      Just (_ :: Path Rel Dir) -> curDirNormalizedFP
+      Nothing -> error "Bug: unexpected empty filepath value"
```
Comment (chrisdone) [inline src/Path.hs]: This is unacceptable. 😭

File: src/Path.hs [hunk 2]
```diff
@@ -370,9 +381,12 @@ filename (Path l) =
 --
 -- The following properties hold:
 --
+-- @dirname $(mkRelDir ".") == $(mkRelDir ".")@
+--
 -- @dirname (p \<\/> a) == dirname a@
 --
 dirname :: Path b Dir -> Path Rel Dir
+dirname (Path []) = Path []
```
Comment (chrisdone) [inline src/Path.hs]: We've stablished that `dirname` is broken: https://github.com/chrisdone/path/issues/18 I don't think we want to make it broken-er. 🙈

File: src/Path.hs (hunk 1 above)
Comment (chrisdone) [inline src/Path.hs]: Perhaps just test for empty string? The typeable business is a bit worrying.

**Comment by chrisdone** [comment]
I'm not comfortable with `.` in general. 

But I'll think about it, it seems reasonable on the surface.

File: src/Path.hs (hunk 1 above)
Comment (harendra-kumar) [inline src/Path.hs]: Can be avoided. The dynamic type check only lets us catch the bug case which is just a defensive programming. But I am curious to know the reasons for avoiding typeable or dynamic type checks?

BTW, why don't we just ban `toFilePath` and use only the APIs with restricted types?

File: src/Path.hs (hunk 2 above)
Comment (harendra-kumar) [inline src/Path.hs]: Not sure what's wrong here? This is just for `dirname "." = "."`, applies only to the `Path Rel Dir` case. `Path []` will never occur for `Path Abs Dir`.

**Comment by harendra-kumar** [comment]
> I'm not comfortable with . in general.

I can understand that. `.` is a special name and might be a bit overloaded in its meaning. In this particular case it is completely equivalent to an empty relative path or the identity of relative path composition. The utility is to express the empty case in composition. Instead of using a `""` I picked the existing notion of `.` for representing this case because it represents the same thing and we already compose it in exactly the same way (`x, ./x, x/.` represent the same relative path). Plain `""` can also be used for representing it instead of `.`, after all its just an empty relative path.

Think about it. Drop it if you still don't like it.

**Comment by sjakobi** [comment]
I want a `Path Rel Dir` that represents the current directory.

In particular I think that (at least on POSIX) this should hold: 

```
fromRelDir <$> parseRelDir "." == Just "."
```

Can someone check what `System.FilePath.Windows.isValid "."` returns please?

**Comment by chrisdone** [comment]
It's difficult to articulate what makes me uncomfortable about `.` I'll try.

`x </> y` is a semigroup operation and it has no zero as a monoid has. You *either have a path or you don't*, and there's **no normalization of paths**, at all, which is important for reasoning. Why does `.` need normalization? Without `.`, path's literally list appending. `x </> y` is literally `Path x </> Path y = Path (x ++ y)` (because directories are always suffixed with `/`). It doesn't get much simpler than that.

Introducing `x </> "." == x` adds an exception to this easy reasoning rule. 

If we had it so that `"x/" </> "."` produced `"x/."` that might seem reasonable, but then what is `"x/" </> "." </> "."`? `x/..`? Now we're referring to the parent directory. 

Introducing `"x/" </> "./" == "x/./"` adds an exception in the sense that equality no longer works. `x </> "./" == x` doesn't hold, even though semantically they **are the same path**. This messes with programmer's logic of their internal functions. Think of putting paths in a hashmap, or whatever.

---

I think that `""` empty list perhaps may fare better in retaining reasonability. But there are a bunch of functions (e.g. `rm`) that fail when given an empty string, so what is the value (pun intended) of an empty path?

I might be more +1 on empty paths though. Perhaps it's best to motivate these with "I tried to use path to do X and it didn't work/is a pain in the arse" then you can assess the trade-offs.

@sjakobi A best of both worlds scenario might be to parse `.` as `""`, thereby supporting user input better, and continuing to serve the programmer with good reasoning properties.

**Comment by NorfairKing** [comment]
> A best of both worlds scenario might be to parse . as ""
Only if you render "" back to "."

**Comment by harendra-kumar** [comment]
@chrisdone It seems you are echoing change in the commit. From the change:

```
      -- Represent a "." in relative dir path as "" internally so that it
      -- composes without having to renormalize the path.
```

The change is just making Rel paths as a monoid instead of semigroup by providing an empty path as identity. A "." is parsed as empty string and an empty string is showed as a ".".

**Comment by chrisdone** [comment]
@harendra-kumar :+1: 

That would make it the only thing in Path that renders differently, otherwise `toFilePath` for everything else is `id`. That's a nice property...

**Comment by harendra-kumar** [comment]
@sjakobi need to add the parenthesis as well for precedence parsing

```
>> (fromRelDir <$> parseRelDir ".") == Just "./"
True
```

BTW, this is already added as a test in the PR, in a slightly different form though.

I rebased, rebuilt and uploaded again. You can review the change and let me know if there is something that needs to be taken care of.

**Comment by sjakobi** [comment]
What should we do about the [`Show` instance](https://github.com/commercialhaskell/path/blob/467906da26eaf850abf561c881e8043d2b64f139/src/Path/Internal.hs#L53)? Change it to `show = show . toFilePath`? Honestly I'd prefer a derived `Show` instance but I'm not sure whether it's worth changing it now…

That sentence is already wrong BTW:

> `-- | Same as 'Path.toFilePath'.`

I'll do a more thorough review once tests are running on Windows again, i.e. once at least #39 is fixed.

**Comment by sjakobi** [comment]
In validity-test, [this line](https://github.com/commercialhaskell/path/blob/467906da26eaf850abf561c881e8043d2b64f139/test/Path/Gen.hs#L47) needs to be removed.

**Comment by harendra-kumar** [comment]
> In validity-test, this line needs to be removed.

I have not yet looked at how this is being used, but I am wondering if this works then why didn't the tests fail? When I ran the tests, all tests passed.

**Comment by chrisdone** [comment]
> That sentence is already wrong BTW:

>     -- | Same as 'Path.toFilePath'.

That commit is just out of date https://github.com/commercialhaskell/path/blob/master/src/Path/Internal.hs#L48

**Comment by sjakobi** [comment]
> That commit is just out of date https://github.com/commercialhaskell/path/blob/master/src/Path/Internal.hs#L48

Ah sorry, I must have looked at the Haddocks for v0.5.12.

**Comment by sjakobi** [comment]
> I have not yet looked at how this is being used, but I am wondering if this works then why didn't the tests fail? When I ran the tests, all tests passed.

@harendra-kumar The validity-test testsuite is pretty flaky right now. It probably just didn't come up with `$mkRelDir "."`

File: test/Path/Gen.hs [hunk 3]
```diff
@@ -45,7 +45,6 @@ instance Validity (Path Rel Dir) where
     && FilePath.hasTrailingPathSeparator fp
     && FilePath.isValid fp
     && not (null fp)
-    && fp /= "."
     && fp /= ".."
     && not (".." `isInfixOf` fp)
     && (parseRelDir fp == Just p)
```
Comment (sjakobi) [inline test/Path/Gen.hs]: This line need to be changed too now. Maybe it should even be refactored into a idempotency property:

```
parse (toFilePath p) == Just p
```

File: src/Path/Internal.hs [hunk 4]
```diff
@@ -45,24 +49,38 @@ instance Eq (Path b t) where
...
-  toEncoding (Path x) = toEncoding x
+  toEncoding = toEncoding . toFilePath
   {-# INLINE toEncoding #-}
 #endif
 
 instance Hashable (Path b t) where
-  hashWithSalt n (Path path) = hashWithSalt n path
+  -- hash of "" is just the salt, so use toFilePath to convert
+  -- it to "." instead to get a reasonable hash
+  hashWithSalt n path = hashWithSalt n (toFilePath path)
```
Comment (sjakobi) [inline src/Path/Internal.hs]: Why is this necessary?

File: test/Path/Gen.hs (hunk 3 above)
Comment (harendra-kumar) [inline test/Path/Gen.hs]: Ah, it seems I removed the wrong line! It still cannot be "." but it can be null. So we need to remove the `not (null fp)`  condition and keep the `fp /= "."` condition.

File: src/Path/Internal.hs (hunk 4 above)
Comment (harendra-kumar) [inline src/Path/Internal.hs]: For the empty string case (`""`) this function will evaluate to just `n`, the salt. Not sure if that can create some unexpected problems. I changed this just to be have a hash corresponding to the "." value instead. I do not have a solid reason to back it up whether this is any better. I can remove it if you would like that.

Comment (sjakobi) [inline src/Path/Internal.hs]: Thanks for the explanation about the `""` case ! Would you mind adding it to the comment so future generations can profit too? :smile:

Comment (harendra-kumar) [inline src/Path/Internal.hs]: It seems you missed the explanation which is already there right above the changed line :-) Is that enough or need to be rephrased for more clarity?

Comment (sjakobi) [inline src/Path/Internal.hs]: Strange, I think I had read it but obviously I didn't really get what it said. No need to change it I guess I'm just tired. :/

Comment (harendra-kumar) [inline src/Path/Internal.hs]: Communication is the hardest thing to get right! Let me see if it can be made better.

**Comment by harendra-kumar** [comment]
Not sure why the cabal build is failing. The tests do not show any failures.

Let me know if there are any more comments on this change. Once the PR to fix the validity test is merged I will fix the conflicts and squash the commits.

**Comment by harendra-kumar** [comment]
Also, this is a behavioral change, so will have to be clearly noted in the changelog and the versioning will have to be handled appropriately.

**Comment by harendra-kumar** [comment]
I squashed the commits, once the change is approved by the maintainers, we can add the changelog as well. I will need the version under which the changelog should be added.

@sjakobi the cabal tests are failing on travis. I suspect it might be the "cabal haddock" command which is failing. I do not have a working local cabal setup right now to figure out what's wrong. I will have to try removing the `grep` part from `travis.yml` so that we can see the full output and figure out what's going on.

Stack haddock succeeds.

**Comment by sjakobi** [comment]
> @sjakobi the cabal tests are failing on travis. I suspect it might be the "cabal haddock" command which is failing. I do not have a working local cabal setup right now to figure out what's wrong.

Right, we have this check that for both `Path`, and `Path.Internal`, haddock coverage is 100%. Apparently `curDirNormalizedFP` doesn't have haddocks, so the check fails. Can you simply add a haddock comment there?

@mrkkrp Could you try making the error in case of sub-100% Haddock coverage more clear? Alternatively we could disable the check or at least relax it a bit – 100% coverage in `Path.Internal` doesn't seem _that_ important too me.

**Comment by sjakobi** [comment]
It looks like you'll need to have some (even more!) patience with us here. I'm quite confident in this PR but I don't want to merge it while our test suites aren't working on Windows (see #74 and #39).

**Comment by harendra-kumar** [comment]
No hurry! Do the right thing.

**Comment by harendra-kumar** [comment]
After the `shrinkValidWith` fix in commit 6adaaaa above the validity tests get stuck in a busy loop it seems. Can someone take a look? Seems to be a bug in the validity generator.

cc @NorfairKing

**Comment by NorfairKing** [comment]
Exitfailure -9 is probably due to memory exhaustion, I think.
I don't see why ...

**Comment by NorfairKing** [comment]
I think I found it!
Part of the validity instance is to ensure that the contents of Path match what you get after parsing the path. You need to fix that too :)

**Comment by harendra-kumar** [comment]
Shrink was running in infinite loop because there was no termination condition. I have fixed it now. But now stack tests are failing because of stack download failure.

**Comment by NorfairKing** [comment]
That sounds like a flaky download, you can rerun the build on travis :)

**Comment by harendra-kumar** [comment]
I think only the repository owners can restart the build on travis. The only way I can restart is by pushing another change.

**Comment by NorfairKing** [comment]
Just restarted the build

**Comment by NorfairKing** [comment]
@harendra-kumar No activity in two weeks.
I'd say use best judgement at this point.

It's worth pointing out that you'll break `toFilePath` with this change.

**Comment by harendra-kumar** [comment]
@NorfairKing the maintainers have to decide whether to merge it or not.

Not sure what you mean by breaking `toFilePath`, it has been modified/fixed as part of this change. Did you miss that or do you mean that's not enough?

**Comment by NorfairKing** [comment]
> @NorfairKing the maintainers have to decide whether to merge it or not.

I am one of those, but I'm not comfortable making the decision by myself here.

> Not sure what you mean by breaking toFilePath, it has been modified/fixed as part of this change. Did you miss that or do you mean that's not enough?

I made a wrong assumption: that `/` is also stored as `Path []`. My appologies.

On the other hand, this breaks any workflow where a user has defined their own alternatives to `Abs` or `Rel`, which I'm fine with.

**Comment by mrkkrp** [comment]
> No activity in two weeks.

Sorry guys, quite busy these days, but I'll have a vacation soon, I wanted to return to this then. If you decide on something without me it's OK, as long as it's well thought-out.

**Comment by NorfairKing** [comment]
> I'll have a vacation soon, I wanted to return to this then. 

SGTM. No hurries.

**Comment by mrkkrp** [comment]
OK, I have reviewed the PR and the thread again. I'm +1 on this approach, let's prefer it to #90 . However, it still does not fix #18. With the code from this PR:

```
λ> dir <- parseAbsDir "/"
λ> dirname dir
"/"
λ> :t dirname dir
dirname dir :: Path Rel Dir
```

Because:

```
λ> (FilePath.splitPath "/")
["/"]
λ> last (FilePath.splitPath "/")
"/"
```

I think `dirname dir` should return that `"."` path here, right?

Another thing, @harendra-kumar could you use `""` instead of `[]` when you pattern-match? It's a lot easier for me to read, as I expect paths inside of `Path`, so it should be string literals, not lists. Just a perception thing.

**Comment by mrkkrp** [comment]
I propose we release 0.6.0 after this PR is merged.

**Comment by harendra-kumar** [comment]
> I think dirname dir should return that "." path here, right?

Yes, it was not part of the original change because it was yet to be discussed at the time. I can include it as part of this change now or fix it separately under #18. Which one is preferred?

**Comment by harendra-kumar** [comment]
For future reference, most of the discussion related to this change is in #90.

**Comment by mrkkrp** [comment]
> Which one is preferred?

I think it's best to include the fix here, so we don't keep the project in "half-fixed" state.

**Comment by mrkkrp** [comment]
We also need tests making sure that #18 is fixed here.

**Comment by mrkkrp** [comment]
Let's add a changelog entry for 0.6.0 perhaps? What do others think? Is there anything else preventing this sort of change from being accepted?

@chrisdone Are you OK with the current state of the PR?

**Comment by NorfairKing** [comment]
I like it. I approve of the changelog entry idea as well.

Side-note. What happens if someone wants to call their file '.'?

**Comment by harendra-kumar** [comment]
> Side-note. What happens if someone wants to call their file .?

However, if this library is used to represent paths other than filesystem paths then one may want to use `.` as a component of their path instead of using it as a relative root. But in general you will still need some notion of path separator, absolute root and a relative root of a path. In a more general path library these notions can be abstracted instead of hard coding them. But that will make the library a bit more inconvenient to use.

**Comment by NorfairKing** [comment]
>  . and .. are special directory entries created by the filesystem automatically.

I completely forgot about that, thank you for clearing it up.

I'm all for this change :) :+1:

**Comment by mrkkrp** [comment]
If there is no objections, I'm going merge this.

**Comment by NorfairKing** [comment]
SGTM
